File: esp32/plugins/aqm0802a.ts
modules['@makestack/aqm0802a'] = class AQM0802A {
    public bus: any;

    constructor(address = 0x3e) {
        this.bus = new I2C({ address });

        for (const cmd of [0x39, 0x14, 0x74, 0x56, 0x6c, 0x0c]) {
            this.bus.writeSync([0x00, cmd]);
        }

        this.clear();
    }

    public clear() {
        this.bus.writeSync([0x00, 0x01]);
        Timer.busywait(50);
    }

    public update(text: string) {
        const shortenedText = text.toString().substring(0, 16);

        this.clear();
        for (let i = 0; i < shortenedText.length; i++) {
            this.bus.writeSync([0x00, 0x80 + ((i > 7) ? 0x40 + i - 8 : i)]);
            Timer.busywait(50);
            this.bus.writeSync([0x40, shortenedText[i].charCodeAt(0)]);
            Timer.busywait(50);
            i++;
        }
    }
}

File: plugins/aqm0802a/aqm0802a.cpp
#include "aqm0802a.h"

AQM0802A::AQM0802A(int addr) : addr(addr) {
    command(0x39);
    command(0x14);
    command(0x74);
    command(0x56);
    command(0x6c);
    command(0x0c);
}

void AQM0802A::command(int cmd) {
    Wire.beginTransmission(addr);
    Wire.write(0x00);
    Wire.write(cmd);
    Wire.endTransmission(addr);
    delayMicroseconds(30);
}

void AQM0802A::write(char ch) {
    Wire.beginTransmission(addr);
    Wire.write(0x40);
    Wire.write(ch);
    Wire.endTransmission(addr);
    delayMicroseconds(30);
}

void AQM0802A::clear() {
    command(0x01);

}
void AQM0802A::update(const char *str) {
    clear();
    for (int i = 0; str[i] != '\0'; i++) {
        command(0x80 + ((i >= 8) ? (0x40 + i - 8) : i));
        write(str[i]);
    }
}
